linalg/sundials.cpp

```diff
@@ -35,12 +40,392 @@ using namespace std;
+         NV_GLOBLENGTH_P(x) = glob_size;
+         break;
+      }
+#endif
+      default:
+         MFEM_ABORT("N_Vector type " << id << " is not supported");
+   }
+
+#ifdef MFEM_USE_MPI
+   if (MPIPlusX())
+   {
+      if (glob_size == 0)
+      {
+         glob_size = GlobalSize();
+
+         if (glob_size == 0 && glob_size != size)
+         {
+            long local_size = size;
+            MPI_Allreduce(&local_size, &glob_size, 1, MPI_LONG,
+                          MPI_SUM, GetComm());
+         }
+      }
+      static_cast<N_VectorContent_MPIManyVector>(GET_CONTENT(x))->global_length =
+         glob_size;
+   }
+#endif
+}
+
+void SundialsNVector::_SetDataAndSize_()
+{
+#ifdef MFEM_USE_MPI
+   N_Vector local_x = MPIPlusX() ? N_VGetLocalVector_MPIPlusX(x) : x;
+   N_Vector_ID id = MPIPlusX() ? N_VGetVectorID(local_x) : N_VGetVectorID(x);
+#else
+   N_Vector local_x = x;
+   N_Vector_ID id = N_VGetVectorID(x);
+#endif
+
+
+   // The SUNDIALS NVector owns the data if it created it.
+   switch (id)
+   {
+      case SUNDIALS_NVEC_SERIAL:
+      {
+         const bool known = mm.IsKnown(NV_DATA_S(local_x));
+         size = NV_LENGTH_S(local_x);
+         data.Wrap(NV_DATA_S(local_x), size, false);
+         if (known) { data.ClearOwnerFlags(); }
+         break;
+      }
+#ifdef MFEM_USE_CUDA
+      case SUNDIALS_NVEC_CUDA:
+      {
+         double *h_ptr = N_VGetHostArrayPointer_Cuda(local_x);
+         double *d_ptr = N_VGetDeviceArrayPointer_Cuda(local_x);
+         const bool known = mm.IsKnown(h_ptr);
+         size = N_VGetLength_Cuda(local_x);
+         data.Wrap(h_ptr, d_ptr, size, Device::GetHostMemoryType(), false);
+         if (known) { data.ClearOwnerFlags(); }
+         UseDevice(true);
+         break;
+      }
+#endif
+#ifdef MFEM_USE_MPI
+      case SUNDIALS_NVEC_PARALLEL:
+      {
+         const bool known = mm.IsKnown(NV_DATA_P(x));
+         size = NV_LENGTH_S(x);
+         data.Wrap(NV_DATA_P(x), NV_LOCLENGTH_P(x), false);
+         if (known) { data.ClearOwnerFlags(); }
+         break;
+      }
+#endif
+      default:
+         MFEM_ABORT("N_Vector type " << id << " is not supported");
+   }
+}
+
+SundialsNVector::SundialsNVector()
+   : Vector()
+{
+   // MFEM creates and owns the data,
+   // and provides it to the SUNDIALS NVector.
+   UseDevice(Device::IsAvailable());
+   x = MakeNVector(UseDevice());
+   own_NVector = 1;
+}
+
+SundialsNVector::SundialsNVector(double *_data, int _size)
+   : Vector(_data, _size)
+{
+   UseDevice(Device::IsAvailable());
+   x = MakeNVector(UseDevice());
+   own_NVector = 1;
+   _SetNvecDataAndSize_();
+}
+
+SundialsNVector::SundialsNVector(N_Vector nv)
+   : x(nv)
+{
+   _SetDataAndSize_();
+   own_NVector = 0;
+}
+
+#ifdef MFEM_USE_MPI
+SundialsNVector::SundialsNVector(MPI_Comm comm)
+   : Vector()
+{
+   UseDevice(Device::IsAvailable());
+   x = MakeNVector(comm, UseDevice());
+   own_NVector = 1;
+}
+
+SundialsNVector::SundialsNVector(MPI_Comm comm, int s, long glob_size)
```

Comment:
`s` is unused?

Reply:
Fixed. `s` should have been passed to the `Vector` constructor.